### PR TITLE
[SPARK-49064][BUILD] Upgrade Kafka to 3.8.0

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to upgrade `kafka` from `3.7.1` to `3.8.0`.

### Why are the changes needed?
https://downloads.apache.org/kafka/3.8.0/RELEASE_NOTES.html

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Since the title is changed, could you make it `Ready for review`?

File: connector/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala [hunk 1]
```diff
@@ -117,7 +118,7 @@ class KafkaRDDSuite extends SparkFunSuite {
     )
     messages.foreach { case (k, v) =>
       val record = new SimpleRecord(k.getBytes, v.getBytes)
-      log.appendAsLeader(MemoryRecords.withRecords(CompressionType.NONE, record), 0);
+      log.appendAsLeader(MemoryRecords.withRecords(Compression.NONE, Seq(record): _*), 0);
```
Comment (dongjoon-hyun) [inline connector/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala:121]: This is a new feature of Kafka 3.8.0 (KAFKA-7632, KIP-390).
- https://github.com/apache/kafka/pull/15516

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM for Apache Spark 4.0.0-preview2. Thank you for keeping tracking this area, @panbingkun .

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 2]
```diff
@@ -447,8 +449,9 @@ class KafkaTestUtils(
     sendMessages(msgs.toImmutableArraySeq)
   }
 
+  private val cleanupLogsPrivateMethod = PrivateMethod[LogManager](Symbol("cleanupLogs"))
   def cleanupLogs(): Unit = {
```
Comment (panbingkun) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala:453]: In v3.8.0, the access modifier of method `cleanupLogs` has been changed to `private`

- v3.7.1
https://github.com/apache/kafka/blob/3.7.1/core/src/main/scala/kafka/log/LogManager.scala#L1390
image

- v3.8.0
https://github.com/apache/kafka/blob/3.8.0/core/src/main/scala/kafka/log/LogManager.scala#L1395
image

File: connector/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala [hunk 3]
```diff
@@ -117,7 +118,7 @@ class KafkaRDDSuite extends SparkFunSuite {
...
     }
```
Comment (panbingkun) [inline connector/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala:122]: In v3.8.0, the signature of method `withRecords` has been changed to `MemoryRecords withRecords(Compression compression, ...)`

- v3.7.1
https://github.com/apache/kafka/blob/3.7.1/clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java#L594
image

- v3.8.0
https://github.com/apache/kafka/blob/3.8.0/clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java#L597
image

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 4]
```diff
@@ -447,8 +449,9 @@ class KafkaTestUtils(
...
-    server.logManager.cleanupLogs()
+    server.logManager.invokePrivate(cleanupLogsPrivateMethod())
```
[the remark this answers is not in the document]
Comment (panbingkun) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala:454]: Yea, that's right, it was this PR that caused this change.

**Comment by panbingkun** [comment]
> Since the title is changed, could you make it `Ready for review`?

Done.

**Comment by panbingkun** [comment]
> +1, LGTM for Apache Spark 4.0.0-preview2. Thank you for keeping tracking this area, @panbingkun .

Thank you for your review! ❤️

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
The failed `OracleIntegrationSuite` is irrelevant to this PR. 

Merged to master for Apache Spark 4.0.0-preview2.

**Pull request closed without merging**